Add basic API docs (#128)
JuliaGaussianProcesses/AbstractGPs.jl: docs/src/api.md

# FiniteGPs and AbstractGPs

> There are two core types that AbstractGPs provides: `FiniteGP`, and `AbstractGP`.

Reviewer: It seems a bit strange the the two core types are not exported, isn't it?

Author: I kind of agree. At present, AbstractGPs serves two competing purposes -- it both specifies an API, and provides a few concrete implementations of said API (e.g. `GP`). Users of said concrete implementations shouldn't ever need to interact with either `AbstractGP`s or `FiniteGP`s, just things like `GP`.

If we split this package up to separate the APIs from the implementations, it would probably make sense to export `AbstractGP` and `FiniteGP`.

Perhaps I should comment on this in the docs?

Reviewer: Yes, I guess a short comment similar to your explanation here would be good.

## FiniteGP

AbstractGPs.jl provides the `FiniteGP` type. It represents the multivariate Gaussian induced by "indexing" into a GP `f` at a collection of points `x`, and adding independent zero-mean noise with variance `Σ`:
```julia
fx = f(x, Σ)

# The above is equivalent to:

> fx = AbstractGPs.FiniteGP(f, x, Σ)

Reviewer: Maybe only one construction should be allowed and/or recommended?

Author: Good point. My intention is that the former is preferred -- I've just documented the latter to make it clear what the former is doing. Would a comment along the lines of "This is just for reference -- please use the above syntax" be sufficient?

Reviewer: Yes :+1:

```

The `FiniteGP` has two API levels.

> The first should be supported by all `FiniteGPs`, while the second will only be supported by a subset.

Reviewer: Matter of taste but I prefer to use inline code only for correct Julia code:
```suggestion
The first should be supported by all `FiniteGP`s, while the second will only be supported by a subset.
```

### Primary Public API

These are user-facing methods.
You can expect them to be implemented whenever you encounter a `FiniteGP`.

> If you are building something on top of AbstractGPs.jl, try to implement it in terms of these functions.

Reviewer: Maybe we should consistently write AbstractGPs.jl or AbstractGPs? I don't mind which alternative we use but it would be nice to be consistent.

#### Required Methods

```@docs
rand
marginals
logpdf(::AbstractGPs.FiniteGP, ::AbstractVector{<:Real})
posterior(::AbstractGPs.FiniteGP, ::AbstractVector{<:Real})
mean(::AbstractGPs.FiniteGP)
var(::AbstractGPs.FiniteGP)

```

#### Optional methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
mean_and_var(::AbstractGPs.FiniteGP)
```

### Secondary Public API

While the covariance matrix of any multivariate Gaussian is defined, it is not always a good idea to actually compute it.
Fortunately, it's often the case that you're not actually interested in the covariance matrix per-se, rather the other quantities that you might use it to compute (`logpdf`, `rand`, `posterior`).
This is similar to the well-known observation that you rarely need the inverse of a matrix, you just need to compute the inverse multiplied by something, so it's considered good practice to avoid ever explicitly computing the inverse of a matrix so as to avoid the numerical issues associated with it.

> This is important, for example, as `TemporalGPs.jl` is able to implement all of the Primary Public API in linear time in the dimension of the `FiniteGP`, as it never needs to evaluate the covariance matrix.

Reviewer: Maybe add a link to TemporalGPs? Another slight inconsistency here is that you write `TemporalGPs.jl` (as inline code) but used AbstractGPs and AbstractGPs.jl before (as regular text).

However, for many (probably the majority of) GPs, this acceleration isn't possible, and there is really nothing lost by explicitly evaluating the covariance matrix.
We call this the Secondary Public API, because it's available a large proportion of the time, but should be avoided if at all possible.

#### Required Methods

```@docs
cov(::AbstractGPs.FiniteGP)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
mean_and_cov(::AbstractGPs.FiniteGP)
```

## Internal AbstractGPs API

> This functionality is not intended to be used directly be users, or those building functionality on top of `AbstractGP` -- they should interact with Primary Public API above, and the Seconary Public API if truly necessary.

Reviewer: ```suggestion
This functionality is not intended to be used directly be users, or those building functionality on top of `AbstractGP` -- they should interact with Primary Public API above, and the Secondary Public API if truly necessary.
```

Reviewer: And maybe one could insert links to the two API sections above?

The reason for this is that some `AbstractGP`s will not actually implement any of these methods, but they will ensure that the Primary Public API is implemented for `FiniteGP`s containing them.
See the next section for more info on this.

Implementing the following API for your own GP type automatically implements both the Primary and Secondary public APIs above in terms of them.

#### Required Methods

```@docs
mean(::AbstractGPs.AbstractGP, ::AbstractVector)
cov(::AbstractGPs.AbstractGP, ::AbstractVector, ::AbstractVector)
var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

> Note that, while we _could_ provide a default implementation for `var` as `diag(cov(f, x))`, this is generally such an inefficient fallback, that we view find it preferable to error if it's not implemented than to ever hit a fallback.

Reviewer: ```suggestion
Note that, while we _could_ provide a default implementation for `var(f, x)` as `diag(cov(f, x))`, this is generally such an inefficient fallback, that we find it preferable to error if it's not implemented than to ever hit a fallback.
```

## When to implement FiniteGP methods

If you have a new subtype of `AbstractGP` and it implements the API above (i.e. you don't mind computing covariance matrices), then you'll not usually need to add new methods involving your own `FiniteGP` -- the fallback implementations will often be completely fine.

> If, on the other hand, you don't want to implement the Internal AbstractGPs API for e.g. performance reasons, then you'll need to implement the Primary Public API directly.

Reviewer: Again maybe one could link to the corresponding sections?

> This is the case in `TemporalGPs.jl` -- the covariance matrix is never actually needed, so we neglect to provide any implementations involving it, instead implementing specialised methods for the Primary Public API.

Reviewer: Maybe some links to TemporalGPs and the API? And maybe don't use inline code for TemporalGPs.jl?

# FiniteGPs and AbstractGPs

> There are two core abstract types that AbstractGPs provides: `FiniteGP`, and `AbstractGP`.

Author: Good point, oversight on my part. Will fix.

> There are two core abstract types that AbstractGPs provides: `FiniteGP`, and `AbstractGP`.

Author: Added. Let me know if you think it's clearer now.

Note that neither of these abstract types are exported as this package also provides user-facing concrete types which are generally what users should interact with.
Package developers, and anyone writing code that is intended to work with any GP in the ecosystem, should work with these types.

## FiniteGP

AbstractGPs provides the `FiniteGP` type. It represents the multivariate Gaussian induced by "indexing" into a GP `f` at a collection of points `x`, and adding independent zero-mean noise with variance `Σ`:
```julia
fx = f(x, Σ)

# The code below is equivalent to the above, and is just for reference.
# When writing code, prefer the above syntax.
fx = AbstractGPs.FiniteGP(f, x, Σ)
```

The `FiniteGP` has two API levels.
The first should be supported by all `FiniteGP`s, while the second will only be supported by a subset.

> ### Primary Public API

Author: I went with the names "Primary" and "Secondary" for lack of better terms. Perhaps "FiniteGP API" and "Extended FiniteGP API" would make things clearer? i.e. all `FiniteGP`s implement the "FiniteGP API", but only some implement the "Extended FiniteGP API", and whether or not the "Extended FiniteGP API" is implemented depends on the `AbstractGP` contained inside?

> ### Primary Public API

Reviewer 2: Yeah I see your point and realised what was meant in my general review comment. This should in any case be properly explained.
In this sense I would probably even start by the `AbstractGP` interface before the `FiniteGP` one.

> ### Primary Public API

Author: > In this sense I would probably even start by the AbstractGP interface before the FiniteGP one.

Yeah, this is where it starts to get slightly tricky, so I think we should think a bit about how to make this work.

The reason that I started with the `FiniteGP` API is because it's the thing that you can use to compute stuff. It turns out that an `AbstractGP` on its own is not especially useful, it's only when you bring in some inputs `x` that you can start to compute things. So in that sense, if you've defined a new `AbstractGP`, the thing that you really want implemented is the FiniteGP API for `FiniteGP{<:MyNewAbstractGP}`. Perhaps I should lead with that?

> ### Primary Public API

Reviewer 2: I think starting with a very concrete example would be quite self-explanatory. For example showing that for `AbstractGP` one would implement `mean(f, x)` but for `FiniteGP` it's by default `mean(f.gp, f.x)` or one can overload it.

> ### Primary Public API

Author: Hmmm yes, something like this seems to me like it would aid clarity. Where in the structure do you think that it should go?

> ### Primary Public API

Reviewer 2: I would put it after line 16-17 where the difference between the two API is described?

> ### Primary Public API

Author: Added. Please resolve if you're happy.

This is important, for example, as [TemporalGPs.jl](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl) is able to [implement](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/blob/master/src/gp/lti_sde.jl) all of the [Primary Public API](@ref) in linear time in the dimension of the `FiniteGP`, as it never needs to evaluate the covariance matrix.

However, for many (probably the majority of) GPs, this acceleration isn't possible, and there is really nothing lost by explicitly evaluating the covariance matrix.
We call this the [Secondary Public API](@ref), because it's available a large proportion of the time, but should be avoided if at all possible.

#### Required Methods

```@docs
cov(::AbstractGPs.FiniteGP)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
mean_and_cov(::AbstractGPs.FiniteGP)
```

> ## Internal AbstractGPs API

Author: This is internal from a user's perspective, and by user I mean someone who is trying to use GPs to model something as opposed to someone who is implementing the API.

I think this is one instance where having both implementations and API in the same package is starting to cause confusion.

> ## Internal AbstractGPs API

Author: @theogf can I mark this as resolved?

> ## Internal AbstractGPs API

Reviewer 2: Yes sorry

> 

Author: Would you like an explanation of what an AbstractGP is here as well, or is the one added at the top sufficient?

> 

Reviewer 2: I think the one you added at the top is now sufficient. It cannot hurt to have it twice though :)

> 

Author: Added. Please let me know if resolved.

# FiniteGP and AbstractGP

AbstractGPs provides one main abstract type, the `AbstractGP`.
Instances of subtypes of `AbstractGP` represent Gaussian processes -- collections of jointly-Gaussian random variables, which may be infinite-dimensional.
Gaussian processes only become useful computationally when you specify a finite-dimensional marginal distribution to work with.
It is for this reason that we provide the `FiniteGP` type.
This type comprises an `AbstractGP`, `f`, and an `AbstractVector`, `x`, and represents the multivariate Gaussian distribution over `f` at every element of `x`.
This distribution, being finite-dimensional, is something that can be used to compute useful things.

> For example, `rand(f(x))` generates a sample from the multivariate Gaussian that is `f` at `x`, and `logpdf(f(x), y)` computes the log (marginal) probability of observing the vector `y` under this distribution.

Reviewer 2: The example is confusing cause we introduce `f(x)` without mentioning that this would build a FiniteGP

Reviewer 2: I think this is the last important point which needs to be addressed.

Author: I liked your suggestion, so I've merged it, so resolving this point.

> 

Reviewer 2: I liked our discussion on what's an AbstractGP and what is not, concretely that an AbstractGP is an object we can use as a map. I think it would be great to have a sentence about it

Author: Could you point towards a particular bit of that discussion that you found useful? I assume you're referring to #122 ?

Reviewer 2: Hmm more specifically the last part of this comment : https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/122#issuecomment-812063387

Author: Oh, I think I might have misunderstood your comment there actually.

I'm quite sloppy with my `~` vs `=`, so possibly I've caused some confusion.

I think that an `AbstractGP` should be thought of as a distribution over functions / maps. If you could actually implement `rand(f)`, you would get a real-valued function. (Sometimes you _can_ implement `rand(f)` e.g. when the kernel is degenerate, but generally you can't).

So I think of `f(x)` as the distribution over the function at the finite collection of locations specified in `x`. Hence `rand(f(x))` gives you a vector of real numbers.

Is this in line with how you've been thinking about this?

Reviewer 2: I mean I am in line in your understanding now.

What I think one needs to explain is why `PosteriorGP` is an `AbstractGP`. From the GP formalism it is not immediately obvious! Because the direct thought (at least mine) would be: "ah yeah the posterior! p(f|y) right? For Gaussian regression that's just a MvNormal therefore it's some kind of `FiniteGP`". So we should explain that "well this can be seen as an `AbstractGP` as well cause from this object you can also create a distribution over functions"

Author: Ahh I see. Yes, we should document this better. I'll add something about this tomorrow.

Author: @theogf I've updated the explanation to be in line with the above. Please let me know if it works for you :)

Consequently, if you create a new subtype `MyNewGP` of `AbstractGP`, the methods that you must implement are not those directly involving `MyNewGP`, but rather those involving
```julia
FiniteGP{<:MyNewGP}
```
We provide two ways in which to do this.
The first is to implement methods directly on `Finite{<:MyNewGP}` -- this is detailed in the [FiniteGP APIs](@ref).
The second is to implement some methods directly involving `MyNewGP`, and utilise default `FiniteGP` methods implemented in terms of these -- this is detailed in the [Internal AbstractGPs API](@ref).

The second interface is generally easier to implement, but isn't applicable for all subtypes of `AbstractGP`.
See [Which API should I implement?](@ref) for further discussion.

> ## Intended Audience

Reviewer 2: Should this go before everything else?

Author: Good shout. Done.

This page is intended for developers.
If you are a user, please refer to our other examples.

## FiniteGP APIs

A `FiniteGP` represents the multivariate Gaussian induced by "indexing" into an AbstractGP `f` at an `AbstractVector` of points `x`, and adding independent zero-mean noise with covariance matrix `Σ`:

> ```julia

Reviewer 2: Maybe define `f` first?

Author: Done.

1. [`GaussianProcessProbabilisticProgramme`](https://github.com/JuliaGaussianProcesses/Stheno.jl/blob/b4e2d20f973a0816272fdf07bdd5896a614b99e1/src/gaussian_process_probabilistic_programme.jl#L8)

#### Required Methods

```@docs
mean(::AbstractGPs.AbstractGP, ::AbstractVector)
cov(::AbstractGPs.AbstractGP, ::AbstractVector, ::AbstractVector)
var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

Note that, while we _could_ provide a default implementation for `var(f, x)` as `diag(cov(f, x))`, this is generally such an inefficient fallback, that we find it preferable to error if it's not implemented than to ever hit a fallback.

## Which API should I implement?

To answer this question, you need to need to know whether or not the default implementations of the [FiniteGP APIs](@ref) work for your use case.

> There are a couple of reasons of which we are aware for why this might not be the case -- possibly there are others.

Reviewer 2: This sentence is probably not needed unless we give some concrete reasons 😊

Author: I give the reasons below! I'll clarify that they're reasons for this.

> # FiniteGPs and AbstractGPs

Reviewer 2: ```suggestion
# FiniteGP and AbstractGP
```
I think the plural leads to confusion with the package

AbstractGPs provides one main abstract type, the `AbstractGP`.
Instances of subtypes of `AbstractGP` represent Gaussian processes -- collections of jointly-Gaussian random variables, which may be infinite-dimensional.

> AbstractGPs only become useful computationally when you specify a finite-dimensional marginal distribution to work with.

Reviewer 2: ```suggestion
Gaussian processes only become useful computationally when you specify a finite-dimensional marginal distribution to work with.
```
I am not sure if you meant the type here, or GPs generally

It is for this reason that we provide the `FiniteGP` type.

> Roughly speaking, this type comprises an `AbstractGP` `f` and an `AbstractVector` `x`, and represents the multivariate Gaussian distribution over `f` at `x`.

Reviewer 2: ```suggestion
This type comprises an `AbstractGP` : `f` and an `AbstractVector` : `x`, and represents the multivariate Gaussian distribution over `f` at every element of `x`.
```
Some kind of separation between type and instance is needed.

This distribution, being finite-dimensional, is something that can be used to compute useful things.
For example, `rand(f(x))` generates a sample from the multivariate Gaussian that is `f` at `x`, and `logpdf(f(x), y)` computes the log (marginal) probability of observing the vector `y` under this distribution.

> Consequently, if you create a new `AbstractGP` subtype, say you called it `MyNewGP`, the methods that you must ensure are implemented are not those directly involving `MyNewGP`, but rather

Reviewer 2: ```suggestion
Consequently, if you create a new subtype `MyNewGP` of `AbstractGP`, the methods that you must implement are not those directly involving `MyNewGP`, but rather those involving
```

```julia
FiniteGP{<:MyNewGP}
```

> We provide two ways in which to do this.

Reviewer 2: This is contradicting the sentence before "you must ensure are implemented"

Author: Sorry, could you expand on what's contradictory here?

Reviewer 2: Well here you say that on me "must" implement FiniteGP{<:NewGP} but this is not true since this is just one API approach (one could just implement the NewGP API

Author: Added a clarifying sentence.

The first is to implement methods directly on `Finite{<:MyNewGP}` -- this is detailed in the [FiniteGP APIs](@ref).
The second is to implement some methods directly involving `MyNewGP`, and utilise default `FiniteGP` methods implemented in terms of these -- this is detailed in the [Internal AbstractGPs API](@ref).

The second interface is generally easier to implement, but isn't applicable for all subtypes of `AbstractGP`.
See [Which API should I implement?](@ref) for further discussion.

> Note that neither `AbstractGP` nor `FiniteGP` are exported as this package also provides user-facing concrete types which are generally what users should interact with.

Reviewer 2: Is this really needed?

Author: Probably not. Will remove.

> Package developers, and anyone writing code that is intended to work with any GP in the ecosystem, should import what they need.

Reviewer 2: ```suggestion
Package developers, and anyone writing code that is intended to work with any GP in the ecosystem, should build upon this two types.
```

## Intended Audience

> If you do not plan to implement your own `AbstractGP` subtype because e.g. you just wish to build a model involving a GP, then the useful part of the API for you is the [FiniteGP APIs](@ref).

Reviewer 2: "involving a GP" is extremely unclear in this context.
Since this whole page is aimed at developers, I think we should explicitely say it at the beginning and redirect more standard users to the examples.

Author: Seems reasonable. Will ammend accordingly.

This is important, for example, as [TemporalGPs.jl](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl) is able to [implement](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/blob/master/src/gp/lti_sde.jl) all of the [Primary Public API](@ref) in linear time in the dimension of the `FiniteGP`, as it never needs to evaluate the covariance matrix.

However, for many (probably the majority of) GPs, this acceleration isn't possible, and there is really nothing lost by explicitly evaluating the covariance matrix.
We call this the [Secondary Public API](@ref), because it's available a large proportion of the time, but should be avoided if at all possible.

#### Required Methods

```@docs
cov(::AbstractGPs.FiniteGP)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
mean_and_cov(::AbstractGPs.FiniteGP)
```

## Internal AbstractGPs API

> This functionality is not intended to be used directly by the users, or those building functionality on top of `AbstractGP` -- they should interact with [Primary Public API](@ref).

Reviewer 2: This is weird, did you mean on top of `FiniteGP` ?

Author: Indeed.

1. [`CompositeGP`](https://github.com/JuliaGaussianProcesses/Stheno.jl/blob/b4e2d20f973a0816272fdf07bdd5896a614b99e1/src/composite/composite_gp.jl#L7)
1. [`GaussianProcessProbabilisticProgramme`](https://github.com/JuliaGaussianProcesses/Stheno.jl/blob/b4e2d20f973a0816272fdf07bdd5896a614b99e1/src/gaussian_process_probabilistic_programme.jl#L8)

#### Required Methods

```@docs
mean(::AbstractGPs.AbstractGP, ::AbstractVector)
cov(::AbstractGPs.AbstractGP, ::AbstractVector, ::AbstractVector)
var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

#### Optional Methods
Default implementations are provided for these, but you may wish to specialise for performance.
```@docs
cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_cov(::AbstractGPs.AbstractGP, ::AbstractVector)
mean_and_var(::AbstractGPs.AbstractGP, ::AbstractVector)
```

Note that, while we _could_ provide a default implementation for `var(f, x)` as `diag(cov(f, x))`, this is generally such an inefficient fallback, that we find it preferable to error if it's not implemented than to ever hit a fallback.

> ## Which API should I implement?

Reviewer 2: That's a great addition!

Author: Thanks :)

# FiniteGP and AbstractGP

## Intended Audience

This page is intended for developers.
If you are a user, please refer to our other examples.

## Introduction

AbstractGPs provides one main abstract type, the `AbstractGP`.
Instances of subtypes of `AbstractGP` represent Gaussian processes -- collections of jointly-Gaussian random variables, which may be infinite-dimensional.
Gaussian processes only become useful computationally when you specify a finite-dimensional marginal distribution to work with.
It is for this reason that we provide the `FiniteGP` type.
This type comprises an `AbstractGP`, `f`, and an `AbstractVector`, `x`, and represents the multivariate Gaussian distribution over `f` at every element of `x`.
This distribution, being finite-dimensional, is something that can be used to compute useful things.

> For example, `rand(f(x))` generates a sample from the multivariate Gaussian that is `f` at `x`, and `logpdf(f(x), y)` computes the log (marginal) probability of observing the vector `y` under this distribution.

Reviewer 2: ```suggestion
For example given `fx = f(x)` a `FiniteGP`, `rand(fx)` generates a sample from the multivariate Gaussian that is `f` at `x`, and `logpdf(fx, y)` computes the log (marginal) probability of observing the vector `y` under this distribution.
```
Here is in connection withe comment above
